src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileTargetLibrary.cs

```diff
@@ -8,80 +8,204 @@
 using NuGet.Shared;
 using NuGet.Versioning;
 
+#nullable enable
+
 namespace NuGet.ProjectModel
 {
     public class LockFileTargetLibrary : IEquatable<LockFileTargetLibrary>
     {
-        public string Name { get; set; }
+        #region Property packing
+
+        private readonly record struct PropertyKey(string PropertyName);
+
+        private static readonly PropertyKey DependenciesKey = new(nameof(Dependencies));
+        private static readonly PropertyKey FrameworkAssembliesKey = new(nameof(FrameworkAssemblies));
+        private static readonly PropertyKey FrameworkReferencesKey = new(nameof(FrameworkReferences));
+        private static readonly PropertyKey RuntimeAssembliesKey = new(nameof(RuntimeAssemblies));
+        private static readonly PropertyKey ResourceAssembliesKey = new(nameof(ResourceAssemblies));
+        private static readonly PropertyKey CompileTimeAssembliesKey = new(nameof(CompileTimeAssemblies));
+        private static readonly PropertyKey NativeLibrariesKey = new(nameof(NativeLibraries));
+        private static readonly PropertyKey BuildKey = new(nameof(Build));
+        private static readonly PropertyKey BuildMultiTargetingKey = new(nameof(BuildMultiTargeting));
+        private static readonly PropertyKey ContentFilesKey = new(nameof(ContentFiles));
+        private static readonly PropertyKey RuntimeTargetsKey = new(nameof(RuntimeTargets));
+        private static readonly PropertyKey ToolsAssembliesKey = new(nameof(ToolsAssemblies));
+        private static readonly PropertyKey EmbedAssembliesKey = new(nameof(EmbedAssemblies));
+        private static readonly PropertyKey PackageTypeKey = new(nameof(PackageType));
+
+        private readonly Dictionary<PropertyKey, object> _propertyValues = new();
+
+        private bool _isFrozen;
+
+        private IList<T> GetListProperty<T>(PropertyKey key)
+        {
+            if (!_propertyValues.TryGetValue(key, out object? value))
+            {
+                if (_isFrozen)
+                {
+                    return Array.Empty<T>();
+                }
+
+                var list = new List<T>();
+                _propertyValues[key] = list;
+                return list;
+            }
 
-        public string Framework { get; set; }
+            return (IList<T>)value;
+        }
 
-        public NuGetVersion Version { get; set; }
+        private void SetListProperty<T>(PropertyKey key, IList<T> list)
+        {
+            System.Diagnostics.Debug.Assert(!_isFrozen, "Attempting to set a property on a frozen instance.");
+            _propertyValues[key] = list;
+        }
 
-        public string Type { get; set; }
+        public void Freeze()
```

Comment:
I have considered this concept for the `PackageSpec` type all up, but I hadn't seen it adopted much. 
Have you seen examples without some of the VS/dotnet repos of this?

Reply:
The closest pattern to this that I've seen is [dependency properties](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/dependency-properties-overview?view=netframeworkdesktop-4.8) in WPF and [`Freezable`](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/freezable-objects-overview?view=netframeworkdesktop-4.8) objects.

Reply:
`PackageSpec` does look like a good candidate for this pattern as well.